packages/rocketchat-otr/client/rocketchat.otr.room.js

```diff
@@ -90,155 +73,115 @@ RocketChat.OTR.Room = class {
-		return window.crypto.subtle.encrypt({
-				name: "RSA-OAEP",
-			},
-			publicKey,
-			new TextEncoder("UTF-8").encode(message) //ArrayBuffer of data you want to encrypt
-		)
-		.then((encrypted) => {
-			return this.bytesToHexString(encrypted);
+	generateKeyPair(refreshKeys) {
+		// Generate an ephemeral key pair.
+		return window.crypto.subtle.generateKey({
+			name: 'ECDH',
+			namedCurve: 'P-256'
+		}, false, ['deriveKey', 'deriveBits']).then((keyPair) => {
+			this.keyPair = keyPair;
+			return crypto.subtle.exportKey('spki', keyPair.publicKey);
+		})
+		.then((exportedPublicKey) => {
+			this.exportedPublicKey = exportedPublicKey;
 		})
 		.catch((err) => {
-			console.log(err);
-			return message;
+			console.error(err);
 		});
 	}
 
-	decryptRSA(message) {
-		return window.crypto.subtle.decrypt({
-				name: "RSA-OAEP",
-			},
-			this.privateKey,
-			new this.hexStringToUint8Array(message) //ArrayBuffer of the data
-		)
-		.then((decrypted) => {
-			//returns an ArrayBuffer containing the decrypted data
-			return new TextDecoder("UTF-8").decode(new Uint8Array(decrypted));
-		})
-		.catch((err) => {
-			console.log(err);
-			return message;
+	importPublicKey(publicKey) {
+		return window.crypto.subtle.importKey('spki', this.hexStringToUint8Array(publicKey), {
+			name: 'ECDH',
+			namedCurve: 'P-256'
+		}, false, []).then((peerPublicKey) => {
+			return crypto.subtle.deriveBits({
+				name: 'ECDH',
+				namedCurve: 'P-256',
+				public: peerPublicKey
+			}, this.keyPair.privateKey, 256);
+		}).then((bits) => {
+			return crypto.subtle.digest({
+				name: 'SHA-256'
+			}, bits);
+		}).then((hashedBits) => {
+			// We truncate the hash to 128 bits.
+			var sessionKeyData = new Uint8Array(hashedBits).slice(0, 16);
+			return crypto.subtle.importKey('raw', sessionKeyData, {
+				name: 'AES-GCM'
+			}, false, ['encrypt', 'decrypt']);
+		}).then((sessionKey) => {
+			// Session key available.
+			this.sessionKey = sessionKey;
 		});
 	}
 
-	getSharedSecret() {
-		return window.crypto.subtle.generateKey({
-				name: "AES-CBC",
-				length: 256, //can be  128, 192, or 256
-			},
-			true, //whether the key is extractable (i.e. can be used in exportKey)
-			["encrypt", "decrypt"] //can be "encrypt", "decrypt", "wrapKey", or "unwrapKey"
-		)
-		.then((sharedSecret) => {
-			this.sharedSecret = sharedSecret;
-			// export public key
-			return window.crypto.subtle.exportKey(
-				"jwk", //can be "jwk" (public or private), "spki" (public only), or "pkcs8" (private only)
-				sharedSecret //can be a publicKey or privateKey, as long as extractable was true
-			)
-			.then((sharedSecretJWK) => {
-				this.sharedSecretJWK = sharedSecretJWK;
-			})
-		})
-	}
+	encrypt(message) {
+		// var clearText = new Uint8Array(message);
+		var clearText = new TextEncoder("UTF-8").encode(message);
+		var userId = new TextEncoder("UTF-8").encode(this.userId);
+		this.serial++;
+		var data = new Uint8Array(1 + 1 + userId.length + clearText.length);
```

Comment:
See, you can have your own format, like we have now. Where you have (length, data) tuples concatenated. Or, you could have `var data = new TextEncoder("UTF-8").encode(EJSON.stringify({serial: serial, clearText: clearText}})`. It would not be so compact, but it would be potentially more future proof. So easier to have various versions of the protocol communicate with each other.